src/main/java/edu/harvard/iq/dataverse/marketplace/controller/api/AuthController.java

```diff
@@ -0,0 +1,260 @@
+package edu.harvard.iq.dataverse.marketplace.controller.api;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import jakarta.validation.Valid;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.web.bind.annotation.CrossOrigin;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import edu.harvard.iq.dataverse.marketplace.model.Role;
+import edu.harvard.iq.dataverse.marketplace.model.User;
+import edu.harvard.iq.dataverse.marketplace.openapi.annotations.AuthAPIDocs;
+import edu.harvard.iq.dataverse.marketplace.payload.ServerMessageResponse;
+import edu.harvard.iq.dataverse.marketplace.payload.auth.RoleDTO;
+import edu.harvard.iq.dataverse.marketplace.payload.auth.request.LoginRequest;
+import edu.harvard.iq.dataverse.marketplace.payload.auth.request.RoleCreationRequest;
+import edu.harvard.iq.dataverse.marketplace.payload.auth.request.SignupRequest;
+import edu.harvard.iq.dataverse.marketplace.payload.auth.response.JwtResponse;
+import edu.harvard.iq.dataverse.marketplace.payload.auth.response.RoleCreationResponse;
+import edu.harvard.iq.dataverse.marketplace.repository.RoleRepo;
+import edu.harvard.iq.dataverse.marketplace.repository.UserRepo;
+import edu.harvard.iq.dataverse.marketplace.security.UserDetailsImpl;
+import edu.harvard.iq.dataverse.marketplace.security.jwt.JwtUtils;
+
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+
+
+/**
+ * This class is used to authenticate users and register new users.
+ */
+@CrossOrigin(origins = "*", maxAge = 3600)
+@RestController
+@RequestMapping("/api/auth")
```

Comment:
I said I'd think more about the API paths. It easier to keep everything under "/api/auth" but maybe users should be listed under "/api/users"? 🤔